packages/salesforcedx-vscode-core/src/commands/auth/forceAuthWebLogin.ts

```diff
@@ -22,39 +24,136 @@ import {
   SfdxCommandlet,
   SfdxCommandletExecutor,
   SfdxWorkspaceChecker
 } from '../util';
+import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
 import { ForceAuthLogoutAll } from './forceAuthLogout';
 
-import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
+interface DeviceCodeResponse {
+  user_code: string;
+  device_code: string;
+  interval: number;
+  verification_uri: string;
+}
 
-export class ForceAuthWebLoginExecutor extends SfdxCommandletExecutor<
+export class ForceAuthWebLoginContainerExecutor extends SfdxCommandletExecutor<
   AuthParams
 > {
-  protected showChannelOutput = isSFDXContainerMode();
+  protected showChannelOutput = false;
 
   public build(data: AuthParams): Command {
     const command = new SfdxCommandBuilder().withDescription(
       nls.localize('force_auth_web_login_authorize_org_text')
     );
-    if (isSFDXContainerMode()) {
-      command
-        .withArg('force:auth:device:login')
-        .withLogName('force_auth_device_login');
-    } else {
-      command
-        .withArg('force:auth:web:login')
-        .withLogName('force_auth_web_login');
+
+    command
+      .withArg('force:auth:device:login')
+      .withLogName('force_auth_device_login')
+      .withFlag('--setalias', data.alias)
+      .withFlag('--instanceurl', data.loginUrl)
+      .withArg('--setdefaultusername')
+      .withJson();
+
+    return command.build();
+  }
+
+  public execute(response: ContinueResponse<AuthParams>): void {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+    let deviceCodeReceived = false;
+
+    channelService.streamCommandStartStop(execution);
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(cliResponse => {
+      stdOut += cliResponse.toString();
+
+      if (!deviceCodeReceived) {
+        const authUrl = this.parseAuthUrlFromStdOut(stdOut);
+
+        if (authUrl) {
+          deviceCodeReceived = true;
+          // open the default browser
+          vscode.env.openExternal(vscode.Uri.parse(authUrl, true));
+        }
+      }
+    });
+
+    execution.processExitSubject.subscribe(() => {
+      this.logMetric(execution.command.logName, startTime);
+    });
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private parseAuthUrlFromStdOut(stdOut: string): string | undefined {
+    let authUrl;
+    try {
+      const response = JSON.parse(stdOut) as DeviceCodeResponse;
+      const verificationUrl = response.verification_uri;
```

Comment:
nit: kind of seems like a stretch, but is there any validation we should do on the URL here?  Could someone get a suspect URL to open in the browser if they do bad things to SFDX?

Reply:
I could try to string match parts of the URL we expect but I feel like we are really just a consumer of the CLI and SF OAuth, which I hope we could trust.

If an actor wanted to do something I think they would have to change the code here and get someone to install the new vsix, or man in the middle the cli response or SF auth and send back a new URL in the JSON.

Just let me know what you're comfortable with? =)

Reply:
Yeah def a stretch.  For the desktop vscode user we have them install sfdx as a pre-req so it wouldn't be too far of a stretch for them to install the wrong thing from a bad actor, but in code builder everything is installed for the end user so probably not as much of a concern.